tests/unit/test_client.py

```diff
@@ -1323,6 +1321,50 @@ def return_mock_tuple(**kwargs):
         self.assertFalse(self.endpoint.make_request.called)
 
 
+class TestClientErrors(TestAutoGeneratedClient):
+    def assert_error_response(self, error_response, error_class_name):
+        self.endpoint.make_request.return_value = (
+            mock.Mock(status_code=400), error_response)
+
+        creator = self.create_client_creator()
+
+        service_client = creator.create_client(
+            'myservice', 'us-west-2', credentials=self.credentials)
+
+        error_class = getattr(service_client.meta.exceptions, error_class_name)
+        with self.assertRaises(error_class):
+            service_client.test_operation(Foo='one', Bar='two')
+
+    def test_client_makes_call_with_error(self):
+        error_data = {'Error': {'Code': 'code', 'Message': 'error occurred'}}
+        self.assert_error_response(error_data, 'ClientError')
+        self.assert_error_response(error_data, 'code')
+
+    def test_error_with_no_wire_code(self):
+        error_data = {'Error': {'Code': '404', 'Message': 'Not Found'}}
+        self.assert_error_response(error_data, 'ClientError')
```

Comment:
So for these tests, they should always be ``ClientError`` as any generated exception subclasses from ``ClientError``. I think it would be great to make assertions that differentiates what the exact class are as well such as asserting the name of the class of the exception that is thrown. For example, when I print out the name of the class of this exception, I get ``404`` when it should be client exception.

So I could see always making the assertion the exception thrown is of ``ClientError`` and then making an additional assertion that name of the class is what we except whether it is a specific one or it did not resolve to anything, making it a general ``ClientError``.